packages/runtime/test/integration/close-clear-ssc.ts

```diff
@@ -0,0 +1,131 @@
+    await runtime.optInToApp(john.address, appId, {}, {}, program); // opt-in to app (set new local state)
+
+    // execute clostOut txn
+    await runtime.executeTx(closeOutParams, program, []);
+
+    syncAccount();
+
+    // verify app is deleted from local state
+    const localApp = john.getAppFromLocal(appId);
+    assert.isUndefined(localApp);
+
+    // verify app is NOT deleted from global state
+    const globalApp = runtime.getApp(appId);
+    assert.isDefined(globalApp);
+
+    // since app is deleted from local, local state should be undefined
+    const localVal = runtime.getLocalState(appId, john.address, 'local-key');
+    assert.isUndefined(localVal);
+
+    // since app is not deleted from global, global state should be updated by smart contract
+    const globalVal = runtime.getGlobalState(appId, 'global-key');
+    assert.deepEqual(globalVal, stringToBytes('global-val'));
+  });
+
+  it("should not delete application on CloseOut call if logic is rejected", async function () {
+    // create app
+    const appId = await runtime.addApp(flags, {}, program);
+    closeOutParams.appId = appId;
+    await runtime.optInToApp(john.address, appId, {}, {}, program); // opt-in to app (set new local state)
+
+    // sending txn sender other than creator (john), so txn should be rejected
+    closeOutParams.fromAccount = alice.account;
+
+    await expectTealErrorAsync(
+      async () => runtime.executeTx(closeOutParams, program, []),
+      ERRORS.TEAL.REJECTED_BY_LOGIC
+    );
+
+    // verify app is not deleted from account's local state (as tx is rejected)
+    const res = john.getAppFromLocal(appId);
+    assert.isDefined(res);
+  });
+
+  // clearState call is different from closeOut call as in clear call, app is deleted from account
+  // even if transaction fails
+  it("should delete application on clearState call even if logic is rejected", async function () {
+    // create app
+    const appId = await runtime.addApp(flags, {}, program);
+    const clearAppParams: SSCCallsParam = {
+      type: TransactionType.ClearSSC,
+      sign: SignType.SecretKey,
+      fromAccount: alice.account, // sending txn sender other than creator (john), so txn should be rejected
+      appId: appId,
+      payFlags: {}
+    };
+    await runtime.optInToApp(alice.address, appId, {}, {}, program); // opt-in to app (set new local state)
+    syncAccount();
+
+    // verify before tx execution that local state is present
+    let res = alice.getAppFromLocal(appId);
+    assert.isDefined(res);
+
+    // execute tx
+    await expectTealErrorAsync(
+      async () => runtime.executeTx(clearAppParams, program, []),
+      ERRORS.TEAL.REJECTED_BY_LOGIC
+    );
+
+    // verify app is deleted from account's local state even if tx is rejected after execution
+    res = alice.getAppFromLocal(appId);
+    assert.isUndefined(res);
+
+    // verify global state is not deleted
+    assert.isDefined(runtime.getApp(appId));
+  });
+});
```

Comment:
What happens if user doesn't opt-in for application and tries to closout ?

Reply:
Thanks for suggestion